### ISSUE TITLE
composer 2.0 issue with Pathdownloader 

### ISSUE THREAD

**Issue by mitelg** [composer/composer#7967]
Project structure:
```
<root>
└── mypackage
    └──  mitelg
         └── src
             └──  Mitelg.php
         └── composer.json
└──  vendor
└──  composer.json
```
`composer.json`:

```json
{
    "name": "test/test",
    "license": "MIT",
    "type": "project",
    "repositories": [
        {
            "type": "path",
            "url": "mypackage/*",
            "options": {
                "symlink": true
            }
        }
    ],
    "prefer-stable": true,
    "minimum-stability": "dev",
    "require": {
        "composer/composer": "2.0.x-dev",
        "mypackage/mitelg": "^1.0"
    }
}
```
`mypackage/mitelg/composer.json`
```json
{
    "name": "mypackage/mitelg",
    "version": "1.0.0",
    "autoload": {
        "psr-4": {
            "Mitelg\\": "./src"
        }
    }
}
```
Output of `composer diagnose`:

```
Checking composer.json: WARNING
require.composer/composer : exact version constraints (2.0.x-dev) should be avoided if the package follows semantic versioning
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK
Checking disk free space: OK
Composer version: @package_version@
PHP version: 7.1.26
PHP binary path: /usr/bin/php7.1
```

At first I ran `composer install` and everything is fine. After that I changed the version of mypackage/mitelg in the composer.json to 1.0.2. (Simulating an update of the package via filesystem) Then I run this command to use composer 2.0:

```
./vendor/bin/composer require/update mypackage/mitelg -vv
```

I get the following output:

```
Using version ^1.0 for mypackage/mitelg
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Dependency resolution completed in 0.002 seconds
Analyzed 114 packages to resolve dependencies
Analyzed 240 rules to resolve dependencies
Package operations: 0 installs, 1 update, 0 removals
Updates: mypackage/mitelg:1.0.2

In PathDownloader.php line 57:
                                                                                                                                                               
  [RuntimeException]                                                                                                                                           
  Package mypackage/mitelg cannot install to "/home/mt/www/composer-test/mypackage/mitelg" inside its source at "/home/mt/www/composer-test/mypackage/mitelg"  
                                                                                                                                                               

Exception trace:
 () at /home/mt/www/composer-test/vendor/composer/composer/src/Composer/Downloader/PathDownloader.php:57
 Composer\Downloader\PathDownloader->download() at /home/mt/www/composer-test/vendor/composer/composer/src/Composer/Downloader/DownloadManager.php:217
 Composer\Downloader\DownloadManager->Composer\Downloader\{closure}() at /home/mt/www/composer-test/vendor/composer/composer/src/Composer/Downloader/DownloadManager.php:232
 Composer\Downloader\DownloadManager->download() at /home/mt/www/composer-test/vendor/composer/composer/src/Composer/Installer/LibraryInstaller.php:93
 Composer\Installer\LibraryInstaller->download() at /home/mt/www/composer-test/vendor/composer/composer/src/Composer/Installer/InstallationManager.php:177
 Composer\Installer\InstallationManager->execute() at /home/mt/www/composer-test/vendor/composer/composer/src/Composer/Installer.php:599
 Composer\Installer->doInstall() at /home/mt/www/composer-test/vendor/composer/composer/src/Composer/Installer.php:230
 Composer\Installer->run() at /home/mt/www/composer-test/vendor/composer/composer/src/Composer/Command/RequireCommand.php:195
 Composer\Command\RequireCommand->execute() at /home/mt/www/composer-test/vendor/symfony/console/Command/Command.php:255
 Symfony\Component\Console\Command\Command->run() at /home/mt/www/composer-test/vendor/symfony/console/Application.php:901
 Symfony\Component\Console\Application->doRunCommand() at /home/mt/www/composer-test/vendor/symfony/console/Application.php:262
 Symfony\Component\Console\Application->doRun() at /home/mt/www/composer-test/vendor/composer/composer/src/Composer/Console/Application.php:273
 Composer\Console\Application->doRun() at /home/mt/www/composer-test/vendor/symfony/console/Application.php:145
 Symfony\Component\Console\Application->run() at /home/mt/www/composer-test/vendor/composer/composer/src/Composer/Console/Application.php:106
 Composer\Console\Application->run() at /home/mt/www/composer-test/vendor/composer/composer/bin/composer:57

require [--dev] [--prefer-source] [--prefer-dist] [--no-progress] [--no-suggest] [--no-update] [--no-scripts] [--update-no-dev] [--update-with-dependencies] [--update-with-all-dependencies] [--ignore-platform-reqs] [--prefer-stable] [--prefer-lowest] [--sort-packages] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--apcu-autoloader] [--] [<packages>...]
```

And I expected this to happen:
The package should be updated, but somehow the symlink gets interpreted differently on Version 2.0 of composer. 
I put following code on Composer/Downloader/PathDownloader.php:52
```
var_dump($url);
var_dump(file_exists($url));
var_dump(realpath($url));
var_dump($path);
var_dump(file_exists($path));
var_dump(realpath($path));
```
Output with composer 2.0 is
```
string(16) "mypackage/mitelg"
bool(true)
string(43) "/home/mt/www/composer-test/mypackage/mitelg"
string(50) "/home/mt/www/composer-test/vendor/mypackage/mitelg"
bool(true)
string(43) "/home/mt/www/composer-test/mypackage/mitelg"
```
Output with current stable composer version 1.8.3 is:
```
string(16) "mypackage/mitelg"
bool(true)
string(43) "/home/mt/www/composer-test/mypackage/mitelg"
string(50) "/home/mt/www/composer-test/vendor/mypackage/mitelg"
bool(false)
bool(false)
```

**Comment by omz13** [comment]
I ran into this issue too (with composer 1.8.4 btw), with a plugin that I developed for [kirby](https://github.com/getkirby/kirby)

Essentially, its when you have package in a repo that has `"symlink": true`. When the package is installed, everything is good (composer installs by making a symlink to the package's repo folder). The issue is on `composer update`, because, composer tries to update, but `PathDownloader.php` throws an error because its trying to install into itself.

I think that logically, what needs to happen is something like: If the destination folder is a symlink'd folder and we are about to install a symlink repo, remove the destination folder symlink, otherwise do the overwrite test as before at  https://github.com/composer/composer/blob/522ea033a3c6e72d72954f7cd019a3b75e28f391/src/Composer/Downloader/PathDownloader.php#L52

**Comment by fixpunkt** [comment]
So I've done some digging with @mitelg. 

### What exactly is broken?

While @mitelg's example project contains a path repository below the project root, all subsequent updates of packages symlinked from a path repository anywhere on the filesystem fail with the `'Package %s cannot install to "%s" inside its source at "%s"'` message, i.e. `path` repositories with `symlink: true` are broken in general and cannot be updated when using the 2.0 branch. 😱 

### Why this works in 1.8

When doing `composer update` in 1.8, composer would first remove the symlink from `vendor/x/y` before "downloading" the package from the repo and running into the check in `\Composer\Downloader\PathDownloader::download()`: https://github.com/composer/composer/blob/dd40d74bf6041ddcefa1ebe46fb9117d031b1ccd/src/Composer/Downloader/FileDownloader.php#L224-L225

Since the symlink no longer exists at this point in time, `realpath($path)` returns false in https://github.com/composer/composer/blob/master/src/Composer/Downloader/PathDownloader.php#L52, allowing the update process to continue without hitting the `'Package %s cannot install to "%s" inside its source at "%s"'` check.

### Why this breaks on the `2.0` branch

When updating a path dependency using Composer 2, "downloading" is performed first, before the actual update process is started: https://github.com/composer/composer/blob/0317199507680bf41baaa4b88246515f46ec4e32/src/Composer/Installer/InstallationManager.php#L173-L184. Since in this model, the symlink will only be removed after PathDownloader::download() is called (just as in 1.8, removing the symlink happens in `FileDownloader::update()`, which is called as part of the method called in L184 here), the check in `PathDownloader::download()` fails the install. Also note that in 2.0, the method doesn't do anything anymore except perform the aforementioned check: https://github.com/composer/composer/blob/0317199507680bf41baaa4b88246515f46ec4e32/src/Composer/Downloader/PathDownloader.php#L40-L64

### What can we do about this?

Unfortunately, I'm not really familiar with the composer sources and I'm also not sure how a fix for this would fit into the Composer 2.0 architecture. I'd also like to avoid the risk of any such PRs wasting anyone's time and effort as described in https://github.com/composer/composer/blob/0317199507680bf41baaa4b88246515f46ec4e32/src/Composer/Downloader/PathDownloader.php#L52-L63. However, if @Seldaek could point me in the right direction, I'd be happy to put in the work.

**Comment by fixpunkt** [comment]
So I've done some digging with @mitelg. 

### What exactly is broken?

While @mitelg's example project contains a path repository below the project root, all subsequent updates of packages symlinked from a path repository anywhere on the filesystem fail with the `'Package %s cannot install to "%s" inside its source at "%s"'` message, i.e. `path` repositories with `symlink: true` are broken in general and cannot be updated when using the 2.0 branch. 😱 

### Why this works in 1.8

When doing `composer update` in 1.8, composer would first remove the symlink from `vendor/x/y` before "downloading" the package from the repo and running into the check in `\Composer\Downloader\PathDownloader::download()`: https://github.com/composer/composer/blob/dd40d74bf6041ddcefa1ebe46fb9117d031b1ccd/src/Composer/Downloader/FileDownloader.php#L224-L225

Since the symlink no longer exists at this point in time, `realpath($path)` returns false in https://github.com/composer/composer/blob/master/src/Composer/Downloader/PathDownloader.php#L52, allowing the update process to continue without hitting the `'Package %s cannot install to "%s" inside its source at "%s"'` check.

### Why this breaks on the `2.0` branch

When updating a path dependency using Composer 2, "downloading" is performed first, before the actual update process is started: https://github.com/composer/composer/blob/0317199507680bf41baaa4b88246515f46ec4e32/src/Composer/Installer/InstallationManager.php#L173-L184. Since in this model, the symlink will only be removed after PathDownloader::download() is called (just as in 1.8, removing the symlink happens in `FileDownloader::update()`, which is called as part of the method called in L184 here), the check in `PathDownloader::download()` fails the install. Also note that in 2.0, the method doesn't do anything anymore except perform the aforementioned check: https://github.com/composer/composer/blob/0317199507680bf41baaa4b88246515f46ec4e32/src/Composer/Downloader/PathDownloader.php#L40-L64

### What can we do about this?

Unfortunately, I'm not really familiar with the composer sources and I'm also not sure how a fix for this would fit into the Composer 2.0 architecture. I'd also like to avoid the risk of any such PRs wasting anyone's time and effort as described in https://github.com/composer/composer/blob/0317199507680bf41baaa4b88246515f46ec4e32/src/Composer/Downloader/PathDownloader.php#L52-L63. However, if @Seldaek could point me in the right direction, I'd be happy to put in the work.

**Comment by Seldaek** [comment]
This was fixed by https://github.com/composer/composer/commit/872604ab36b00acb64dd97691c0ccfc77500f3d8
